File: Other Data Structures_qstackheap/le232_implement_queue_using_stacks.java
public class MyQueue {
    private Stack<Integer> stackin;
    private Stack<Integer> stackout;
    
    /** Initialize your data structure here. */
    public MyQueue() {
        stackin = new Stack<Integer>();
        stackout = new Stack<Integer>();
    }
    
    /** Push element x to the back of queue. */
    public void push(int x) {
        stackin.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    public int pop() {
        if (stackout.size() == 0) {
            stackin2out();
        }
        return stackout.pop();
    }
    
    /** Get the front element. */
    public int peek() {
        if (stackout.size() == 0) {
            stackin2out();
        }
        return stackout.peek();
    }
    
    /** Returns whether the queue is empty. */
    public boolean empty() {
        return (stackin.size() + stackout.size()) == 0;
    }
    
    private void stackin2out(){
        while (!stackin.empty()) {
            stackout.push(stackin.pop());
        }
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: Other Data Structures_qstackheap/le232_implementQueueUsingStacks.py
"""
Implement the following operations of a queue using stacks.
（assume that when there is no elements, we will not pop)
push(x) -- Push element x to the back of queue.
pop() -- Removes the element from in front of queue.
peek() -- Get the front element.
empty() -- Return whether the queue is empty.
Notes:
You must use only standard operations of a stack -- which means only push to top, peek/pop from top, size, and is empty operations are valid.
Depending on your language, stack may not be supported natively. You may simulate a stack by using a list or deque (double-ended queue), as long as you use only standard operations of a stack.
You may assume that all operations are valid (for example, no pop or peek operations will be called on an empty queue).

"""

class MyQueue(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.s1 = []
        self.s2 = []

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: void
        """
        self.s1.append(x)

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        self.peek()
        return self.s2.pop()
            
        

    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
       
        if len(self.s2) == 0:
            self.move2s2()
        return self.s2[-1]
    
    def move2s2(self):
        while len(self.s1) != 0:
            self.s2.append(self.s1.pop())
        

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return len(self.s1) + len(self.s2) == 0


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
